qualtran/bloqs/bookkeeping/auto_partition.py

```diff
@@ -0,0 +1,116 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from itertools import chain
+from typing import Dict, Sequence, Set, Tuple
+
+from attrs import evolve, field, frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    QAny,
+    Register,
+    Side,
+    Signature,
+    SoquetT,
+)
+from qualtran.bloqs.bookkeeping.partition import Partition
+from qualtran.cirq_interop.t_complexity_protocol import TComplexity
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class AutoPartition(Bloq):
+    """Wrap a bloq with `Partition` to fit an alternative set of input and output registers
+       such that splits / joins can be avoided in diagrams.
+
+    Args:
+        bloq: The bloq to wrap.
+        partition: A sequence of pairs specifying each register that the wrapped bloq should accept
+            and the registers from `bloq.signature.lefts()` that concatenate to form it.
+        partition_output: If True, the output registers will also follow `partition`.
+            Otherwise, the output registers will follow `bloq.signature.rights()`.
+
+    Registers:
+        [user_spec]: The output registers of the wrapped bloq.
+    """
+
+    bloq: Bloq
+    partition: Sequence[Tuple[Register, Sequence[Register]]] = field(
+        converter=lambda s: tuple((r, tuple(rs)) for r, rs in s)
+    )
```

Comment:
would it be worth having some sort of validation that the list of first entries in this matches bloq.lefts() ?

Reply:
I think the appropriate invariant is that the concatenation of the second entries in `partition` matches `bloq.lefts()`. I'll add this validation.